frontstage/views/passwords.py

```diff
@@ -24,6 +25,46 @@ def forgot_password():
         email_address = request.form.get('email_address')
         post_data = {"email_address": email_address}
 
+        try:
+            token_url = app.config['ONS_TOKEN']
+
+            data = {
+                'grant_type': 'reset_password',
+                'client_id': app.config['RAS_FRONTSTAGE_CLIENT_ID'],
+                'client_secret': app.config['RAS_FRONTSTAGE_CLIENT_SECRET'],
+                'username': email_address,
+            }
+            headers = {
+                'Accept': 'application/json',
+                'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
+            }
+
+            oauth2_response = requests.post(url=token_url, data=data,
+                                            headers=headers, auth=(app.config['RAS_FRONTSTAGE_CLIENT_ID'],
+                                                                   app.config['RAS_FRONTSTAGE_CLIENT_SECRET']))
+            # Check to see that this user account is not locked
+            if oauth2_response.status_code == 401:
+                oauth2Error = json.loads(oauth2_response.text)
+                if 'Unauthorized user credentials' in oauth2Error['detail']:
+                    logger.warning('Email address is not registered')
+                    template_data = {"error": {"type": {"Email address is not registered"}}}
+                    return render_template('passwords/forgot-password.html', _theme='default', form=form,
+                                           data=template_data)
+                elif 'User account locked' in oauth2Error['detail']:
+                    logger.warning('Trying to reset password when user account is locked on the OAuth2 server')
+                else:
+                    logger.error('OAuth 2 server generated 401 which is not understood',
+                                 oauth2error=oauth2Error.get('detail'))
+
+                return render_template('passwords/reset-password.trouble.html', _theme='default',
+                                       data={"error": {"type": "failed"}})
+            if oauth2_response.status_code != 201:
+                logger.error('Unknown error from the OAuth2 server')
+                raise ExternalServiceError(oauth2_response)
+        except requests.ConnectionError as e:
+            logger.warning('Connection error between the server and the OAuth2 service of: {}'.format(exception=str(e)))
```

Comment:
Could of untested lines here and at line 56

Reply:
We could catch requests.ConnectionError in the error handlers. It exposes the url it failed to connect to